### ISSUE TITLE
Latent supervision at time T

### ISSUE THREAD

**Issue by PeidongLi** [BraveGroup/LAW#2]
Thanks for your amazing work! I noted in this paper the best time horizon is 1.5 s, so when in time T,  how do you supervise the latent loss? using `loss = latent_world_model(A_(T-3)) - V_T` ? Does it means you should keep a memory queue with length as 3 in trainning?

**Comment by liyingyanUCAS** [comment]
During training, at time T-3, we use a latent world model to predict the view latent at time T. Then, we extract the actual view latent at time T to serve as supervision. You can load both T-3 and T in the dataloader simultaneously, eliminating the need to maintain a memory queue.

**Comment by liyingyanUCAS** [comment]
Sorry for the late reply, we use one previous frame for temporal aggregation at time T-3.

**Comment by PeidongLi** [comment]
> Sorry for the late reply, we use one previous frame for temporal aggregation at time T-3.

So which frame have grad when training? I have tried to reproduce it but find it hard to converge while both T-3 and T timestamp to calculate loss and grad.

**Comment by liyingyanUCAS** [comment]
It is acceptable to calculate the loss and gradients on both the T-3 and T timestamp frames. However, it is crucial to detach the view latent of the T timestamp when calculating the latent prediction loss, where this latent serves as the supervision for the predicted latent from the T-3 frame.

**Comment by PeidongLi** [comment]
Very thanks for your reply! I'm trying to load two previous frames of current time T for temporal aggregation while T+3 frames as supervision now. Do you load T-4 and T-3 frames for temporal aggregation and T frames for supervision? When will you release the code?

> It is acceptable to calculate the loss and gradients on both the T-3 and T timestamp frames. However, it is crucial to detach the view latent of the T timestamp when calculating the latent prediction loss, where this latent serves as the supervision for the predicted latent from the T-3 frame.

**Comment by turboxin** [comment]
hi @liyingyanUCAS，thanks for sharing this great work, I think the idea is brilliant and am trying to reproduce it. However，after I add latent world model supervision to the model, the latent loss converges to near 0 but waypoint loss diverges. 

Here is my implementation:

- only t-1 and t frame is used
- (t-1) frame is used to predict latent, and is supervised from t  frame latent gt (which is detached when calculating latent loss)
- t frame latent is aggregated by t-1 latent pred and then supervised by waypoint loss
- no torch.zero_grad() is used

Could you please give me some advice on the implementation, thanks a lot!

BTW, when do you plan to release the code, looking forward to it!

**Comment by liyingyanUCAS** [comment]
@turboxin 
''t frame latent is aggregated by t-1 latent pred and then supervised by waypoint loss''
Is this prediction latent generated by the world model or the temporal module? Please ensure that the latent predicted by the world model is not used for temporal aggregation, use the one predicted by the temporal module.

**Comment by liyingyanUCAS** [comment]
@PeidongLi @turboxin Thank you for your attention. We will release the code if we are lucky with the submission.

**Comment by liyingyanUCAS** [comment]
@PeidongLi Please try loading the T-3 frame for temporal aggregation and avoid computing the gradient at the T-3 moment. Then supervise the T frame and the T+3 frame. The T frame is for waypoint loss and latent prediction loss, while the T+3 frame is for waypoint loss.

**Comment by turboxin** [comment]
Hi @liyingyanUCAS , I have a another question, how many modality do you predict for ego planning?

**Comment by PeidongLi** [comment]
> @PeidongLi Please try loading the T-3 frame only instead of T-3 and T-4 frames.

Thanks for your clarification. I.m still curious when latent loss is calculated by` loss_lat = latent_world_model(A_(T-3)) - V_T`, which time should imitation loss be calculated? ` loss_imi = MLP(V_(T-3)) - GT_waypoint(T-3)` or ` loss_imi = MLP(V_(T)) - GT_waypoint(T)`? @liyingyanUCAS

**Comment by liyingyanUCAS** [comment]
> Hi @liyingyanUCAS , thanks for your quick reply and pointing out our mistake!
> 
> I realized that our original implementation was maybe over-simplified and did not obey with your paper, we are trying a new implementation like this, could your please comment on this, thanks again!
> 
> ![law_v2](https://private-user-images.githubusercontent.com/31332811/347331141-e039e2cd-8d81-4a12-9b84-3c48712d9ac8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwOTg3ODQsIm5iZiI6MTcyMTA5ODQ4NCwicGF0aCI6Ii8zMTMzMjgxMS8zNDczMzExNDEtZTAzOWUyY2QtOGQ4MS00YTEyLTliODQtM2M0ODcxMmQ5YWM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE2VDAyNTQ0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmMmQyOTM4MDRmOGJjNTEwMTE4NWU0YTRiMjY4YmYzYzg1NTc4MDY5MGU4NjY0NzljYjhhZDQzZWE5YzFhZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zi80CPBBTkpAgVL8ZZQ5XYf27fqNw1VGMYnB3P0r5wI)

You may apply the waypoint loss at both frame t and frame t+1 to aid in convergence. When computing the latent loss, ensure to detach the gt latent.

**Comment by liyingyanUCAS** [comment]
> Hi @liyingyanUCAS , I have a another question, how many modality do you predict for ego planning?

We use only 1 modality.

**Comment by liyingyanUCAS** [comment]
> > @PeidongLi Please try loading the T-3 frame only instead of T-3 and T-4 frames.
> 
> Thanks for your clarification. I'm still curious when latent loss is calculated by` loss_lat = latent_world_model(A_(T-3)) - V_T`, which time should imitation loss be calculated? ` loss_imi = MLP(V_(T-3)) - GT_waypoint(T-3)` or ` loss_imi = MLP(V_(T)) - GT_waypoint(T)`? or both used? @liyingyanUCAS

Hi, it is both used.
